### DISCUSSION TITLE
Brainstorming Ideas for a Fancy Execute

### DISCUSSION THREAD

**Discussion by c0nn3r** [psycopg/psycopg#64] [Ideas]
Currently, I'm not really satisfied with how any Python SQL libraries handle parameterizing queries. I think it would be interesting to look into alternative syntaxes while psycopg3 is still in beta. I'm thinking something like this:

```python
cur.exec('INSERT INTO students VALUES ({name}, {date})')
```

Everything in a bracket is pulled from the local scope and escaped. If you want to override a variable (for instance, you are using that variable name elsewhere, but would like to use a short name), you can pass it as a named argument to override anything in the local scope:

```python
cur.exec('INSERT INTO students VALUES ({name}, {date})', name=student_name)
```

Depending on how comfortable people are magic, we could support f-string like functionality, sadly [PEP501](https://www.python.org/dev/peps/pep-0501/) is still deferred.

While on first glance I'm sure this looks like a SQL injection waiting to happen, I think this will actually limit SQL injections since it makes the correct approach the easiest.

The other idea that I would like to try is using `typing.get_type_hints` to try to find the correct types for the function parameters. Building off this, I think there could be a `strict=True` argument that raises an exception if the type cannot be identified, this could also potentially integrate with Mypy.

A few ideas for type overrides are as follows. The simplest option is just having the user pass a `typing.TypedDict` to set the types:

```python
cur.exec('INSERT INTO students VALUES ({name}, {date})', name=student_name, types={'name': str})
```

I don't like this option since the type annotation isn't "local" which makes it easy to make a mistake. Another option is using `typing.Annotated`:

```python
cur.exec('INSERT INTO students VALUES ({name}, {date})', name=Annotated[str, student_name])
```

This also isn't a great option, since it requires an import, but really there isn't any reason we couldn't support both. The third option would be to support a type annotation in the actual SQL expression:

```python
cur.exec('INSERT INTO students VALUES ({name: str}, {date})')
```

I like this, but sadly IDEs might find it hard to autocomplete.

Of course, I am more than happy to make a PR for this, I just wanted to get some other opinions before deciding this is useful for other people as well and if it is a good fit for psycopg3.

**Comment by dvarrazzo** [comment]
I would personally love if format strings could be used, such as in:
```python
cur.execute(sql`INSERT INTO students VALUES ({name}, {date})`)
```
but I don't believe that's the direction the language is taking, right?

Observations:

The mechanism for overriding values seems closing the door to pass any option to `execute()`: any keyword we may add in the future may conflict with a parameter name someone has chosen. Maybe there are ways to work the problem around though.

What is the mechanism to extract the variables? would `exec()` (`execute()`?) go up the stack and look for the locals there?

About types hint: it seems a way to infer at import time the types to pass to a query. What would it be useful for? Does it cover cases in which the runtime value gives less information? Types can be specified on the postgres side using a cast (e.g. `%s::text`): this mechanism allow to disambiguate Postgres types which may map to the same Python type. Are there advantages of Python type hints over it?
